MAINT validate parameter in LinearSVM

berkecanrizai opened the pull request:
Reference Issues/PRs
**LinearSVR and LinearSVC** use _validate_parameters from https://github.com/scikit-learn/scikit-learn/issues/23462

What does this implement/fix? Explain your changes.
Under LinearSVR and LinearSVC, a class attribute _parameter_constraints is added. This contains the valid types and values for the parameters of the estimator that are defined under the class.
fit method calls checks parameters at first. Then checks if random_state is a valid one by check_random_state.

Any other comments?
This is my first contribution. Hopefully, the changes are alright.

berkecanrizai committed:
Added _parameter_constraints for both LinearSVC and LinearSVM and added _validate_params() under fit functions of both classes.

berkecanrizai committed:
formatting

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -189,6 +193,22 @@ class LinearSVC(LinearClassifierMixin, SparseCoefMixin, BaseEstimator):
     """
 
+    _parameter_constraints = {
+        "penalty": StrOptions({"l1", "l2"}),
+        "loss": StrOptions({"hinge", "squared_hinge"}),
```
The `StrOptions` need to be part of an iterable list. Same goes for `multi_class`, and `loss` in the validation of `LinearSVR`.
```suggestion
        "penalty": [StrOptions({"l1", "l2"})],
        "loss": [StrOptions({"hinge", "squared_hinge"})],
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -189,6 +193,22 @@ class LinearSVC(LinearClassifierMixin, SparseCoefMixin, BaseEstimator):
+        "penalty": StrOptions({"l1", "l2"}),
+        "loss": StrOptions({"hinge", "squared_hinge"}),
+        "dual": ["boolean"],
+        "tol": [Real],
```
`tol` is a tolerance amount for determining convergence. In this case, it really only is meaningful to have a tolerance value greater than zero, so you'll need to give it actual bounds. For example,

```suggestion
        "tol": [Interval(Real, 0.0, None, closed="left")],
```
where `Interval` is imported from `..utils._param_validation`. Similarly, `C` must be strictly positive and will need a similar construction.

Same applies for `LinearSVR`.

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -189,6 +193,22 @@ class LinearSVC(LinearClassifierMixin, SparseCoefMixin, BaseEstimator):
+        "multi_class": StrOptions({"ovr", "crammer_singer"}),
+        "fit_intercept": ["boolean"],
+        "intercept_scaling": [Real],
+        "class_weight": [None, dict, "balanced"],
```
Even if you want a single string option, e.g. `"balanced"` in this case, you'll still need to use `StrOptions` otherwise the validation function will try to interpret `"balanced"` as a special option such as `"boolean"`.

```suggestion
        "class_weight": [None, dict, StrOptions({"balanced"})],
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -189,6 +193,22 @@ class LinearSVC(LinearClassifierMixin, SparseCoefMixin, BaseEstimator):
+        "intercept_scaling": [Real],
+        "class_weight": [None, dict, "balanced"],
+        "verbose": [Integral],
+        "random_state": [None, Integral, "random_state"],
```
Since `"random_state:"` is such a commonly included parameter for validation, it has its own special helper constraint which is simply `"random_state"`. Again, this doesn't constrain it to only be **the string** `"random_state"` but instead tells the function to process it according to the special helper.

```suggestion
        "random_state": ["random_state"],
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -189,6 +193,22 @@ class LinearSVC(LinearClassifierMixin, SparseCoefMixin, BaseEstimator):
+        "class_weight": [None, dict, "balanced"],
+        "verbose": [Integral],
+        "random_state": [None, Integral, "random_state"],
+        "shrinking": ["boolean"],
```
The tests are hinting that this parameter isn't actually a part of `LinearSVC`

```suggestion
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -242,6 +262,10 @@ def fit(self, X, y, sample_weight=None):
         """
+        self._validate_params()
+
+        check_random_state(self.random_state)
+
```
As mentioned in a prior comment, this doesn't actually validate the random state, we do that by specifying the special `"random_state"` constraint. This builds a random state, and so is often used in e.g. `fit`. Not needed here.

```suggestion
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -242,6 +262,10 @@ def fit(self, X, y, sample_weight=None):
+
+        check_random_state(self.random_state)
+
         if self.C < 0:
             raise ValueError("Penalty term must be positive; got (C=%r)" % self.C)
```
The idea is that with this new centralized validation scheme you can remove any old existing manual validation so long as it is simple like this. If it is more complex, e.g. depends on _the data_ or _other params_ then it can be left in, but this kind of stuff should be removed.
```suggestion
```

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -427,6 +451,19 @@ class LinearSVR(RegressorMixin, LinearModel):
     """
 
+    _parameter_constraints = {
+        "epsilon": [Real],
+        "tol": [Real],
+        "C": [Real],
+        "loss": StrOptions({"epsilon_insensitive", "squared_epsilon_insensitive"}),
+        "fit_intercept": ["boolean"],
+        "intercept_scaling": [Real],
+        "dual": ["boolean"],
+        "verbose": [Integral],
+        "random_state": [None, Integral, "random_state"],
+        "max_iter": [Integral],
```
Same notes as above apply to these.

Micky774 commented on sklearn/svm/_classes.py:
```diff
@@ -476,6 +513,10 @@ def fit(self, X, y, sample_weight=None):
         """
+        self._validate_params()
+
+        check_random_state(self.random_state)
+
```
Same as above, this can be removed

```suggestion
```

Micky774 reviewed (commented):
Hey there @berkecanrizai, thank you for the PR! First off, I'm not a core maintainer so even after my review you'll need the approval of two core maintainers to be merged. Hopefully my review can help you improve the PR a bit and speed up the approval process. Either way, please let me know if you have any questions or concerns.

berkecanrizai committed:
Removed wrong random_state check, removed unnecessary C parameter check, added tol and C strictly positive or 0 to _parameter_constraints. Fixed string option checks in SVR and LinearSvc.

berkecanrizai committed:
Removed LinearSVC and LinearSVR from test_common.py

berkecanrizai commented:
> There also some existing validation tests you'll have to modify or remove, e.g.
> 
> ```
> test_linear_svx_uppercase_loss_penality_raises_error
> test_linearsvc_parameters
> ```

I removed linearSVC and linearSVR from test_common.py. I believe it should be enough but I am not confident with that. Also, thanks for the other suggestions, I have implemented them.

glemaitre committed:
fixes

glemaitre committed:
Merge remote-tracking branch 'origin' into pr/berkecanrizai/23686

glemaitre committed:
fixes

glemaitre committed:
fix import

glemaitre reviewed (approved):
If the CIs passes, LGTM

glemaitre committed:
iter

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/berkecanrizai/23686

jeremiedbb committed:
allow max_iter=0

jeremiedbb commented:
max_iter=0 seems to be valid. I updated the constraints

jeremiedbb reviewed (approved):
LGTM. Thanks @berkecanrizai
